File: src/adjust_entries.rb
#! /usr/bin/env ruby
# coding: utf-8

# Author: UTUMI Hirosi (utuhiro78 at yahoo dot co dot jp)
# License: Apache License, Version 2.0

require 'nkf'

targetfiles = ARGV

if ARGV == []
	puts "Usage: ruby script.rb [FILE]"
	exit
end

filename = ARGV[0]
dicname = filename

file = File.new(filename, "r")
	lines = file.read.split("\n")
file.close

l2 = []
p = 0

lines.length.times do |i|
	s = lines[i].split("	")
	yomi = s[0]
	hyouki = s[4]

	# 表記の全角英数を半角に変換
	hyouki = NKF.nkf("-m0Z1 -W -w", hyouki)

	# 表記の「~」を「〜」に置き換える
	# jawiki-latest-all-titles の表記に合わせる。
	hyouki = hyouki.gsub("~", "〜")

	# 表記の最初が空白の場合は取る
	if hyouki[0] == " "
		hyouki = hyouki[1..-1]
	end

	# 表記の全角カンマを半角に変換
	hyouki = hyouki.gsub("，", ", ")

	# 表記の最後が空白の場合は取る（「，」が「, 」になっている）
	if hyouki[-1] == " "
		hyouki = hyouki[0..-2]
	end

	# 読みにならない文字を削除したhyouki2を作る
	hyouki2 = hyouki.tr(' !?=:・。★☆', '')

	# hyouki2がひらがなとカタカナだけの場合は、読みをhyouki2から作る
	# さいたまスーパーアリーナ
	if hyouki2 == hyouki2.scan(/[ぁ-ゔァ-ヴー]/).join
		yomi = NKF.nkf("--hiragana -w -W", hyouki2)
		yomi = yomi.tr("ゐゑ", "いえ")
	end

	# 読みが2文字以下の場合はスキップ
	if yomi[2] == nil ||
	# hyouki2が1文字の場合はスキップ
	hyouki2[1] == nil ||
	# hyoukiが26文字以上の場合はスキップ
	hyouki[25] != nil ||
	# 読みの文字数がhyouki2の4倍を超える場合はスキップ
	# けやきざかふぉーてぃーしっくす（15文字） 欅坂46（4文字）
	yomi.length > hyouki2.length * 4 ||
	# hyouki2の文字数が読みの文字数より多い場合はスキップ
	# 英数字表記が削除されるのを防ぐため、hyouki2の文字数は (bytesize / 3) とする。
	yomi.length < hyouki2.bytesize / 3 ||
	# 読みがひらがな以外を含む場合はスキップ
	yomi != yomi.scan(/[ぁ-ゔー]/).join ||
	# hyoukiがコードポイントを含む場合はスキップ
	# デコードする場合は次のように行う
	# hyouki = hyouki.gsub(/\\u([\da-fA-F]{4})/){[$1.hex].pack("U")}
	hyouki.index("\\u") != nil ||
	# hyouki2の数字が101以上の場合はスキップ（「100円ショップ」は残す）
	# 国道120号, 3月26日
	hyouki2.scan(/\d/).join.to_i > 100
		next
	end

	l2[p] = yomi + "	" + s[1..3].join("	") + "	" + hyouki
	p = p + 1
end

lines = l2
l2 = []

dicfile = File.new(dicname, "w")
	dicfile.puts lines
dicfile.close

File: src/adjust_entries.py
#!/usr/bin/env python
# coding: utf-8

# Author: UTUMI Hirosi (utuhiro78 at yahoo dot co dot jp)
# License: Apache License, Version 2.0

import jaconv
import re
import sys
from unicodedata import normalize

args = sys.argv[1:]

if not args:
    print("Usage: python script.py [FILE]")
    sys.exit()

file_name = args[0]

with open(file_name, "r", encoding="utf-8") as file:
    lines = file.read().splitlines()

l2 = []

for i in range(len(lines)):
    entry = lines[i].split("\t")
    yomi = entry[0]
    hyouki = entry[4]

    # 表記の全角英数を半角に変換
    hyouki = normalize("NFKC", hyouki)

    # 表記の「~」を「〜」に置き換える
    hyouki = hyouki.replace("~", "〜")

    # 表記の最初が空白の場合は取る
    if hyouki.startswith(" "):
        hyouki = hyouki[1:]

    # 表記の全角カンマを半角に変換
    hyouki = hyouki.replace("，", ", ")

    # 表記の最後が空白の場合は取る（「，」が「, 」になっている）
    if hyouki.endswith(" "):
        hyouki = hyouki[:-1]

    # 読みにならない文字「 !?」などを削除したhyouki2を作る
    hyouki2 = hyouki.translate(str.maketrans('', '', ' .!?-+*=:/・。×★☆'))

    # hyouki2がひらがなとカタカナだけの場合は、読みをhyouki2から作る
    if hyouki2 == ''.join(re.findall('[ぁ-ゔァ-ヴー]', hyouki2)):
        yomi = jaconv.kata2hira(hyouki2)
        yomi = yomi.translate(str.maketrans('ゐゑ', 'いえ'))

    # 読みが2文字以下の場合はスキップ
    # hyouki2が1文字の場合はスキップ
    # hyoukiが26文字以上の場合はスキップ
    # 読みの文字数がhyouki2の4倍を超える場合はスキップ
    # （さくらざかふぉーてぃーしっくす[15文字] 櫻坂46[4文字]」までは残す。）
    # hyouki2の文字数*3が読みのバイト数より小さい場合はスキップ
    # （「あいてぃー[5文字] ITエンジニア[17bytes]」をスキップ。）
    # （「すのーまん[5文字] SNOWMAN[7bytes]」は残す。）
    # 読みがひらがな以外を含む場合はスキップ
    # hyoukiがコードポイントを含む場合はスキップ
    if len(yomi) < 3 or \
            len(hyouki2) < 2 or \
            len(hyouki) > 25 or \
            len(yomi) > len(hyouki2) * 4 or \
            len(yomi) * 3 < len(hyouki2.encode()) or \
            yomi != ''.join(re.findall('[ぁ-ゔー]', yomi)) or \
            '\\u' in hyouki:
        continue

    # hyouki2の数字をつなげると101以上になる場合はスキップ
    # （「国道120号」「3月26日」はスキップ。「100円ショップ」は残す。）
    n = re.sub(r"\D", "", hyouki2)

    if n != "" and int(n) > 100:
        continue

    entry[0] = yomi
    entry[4] = hyouki
    l2.append("\t".join(entry) + "\n")

lines = l2
l2 = []

with open(file_name, "w", encoding="utf-8") as dict_file:
    dict_file.writelines(lines)
